### PR TITLE
[bug] Matrix refactor bug fix: Fix cross scope matrix operations

### PR DESCRIPTION
**Pull request by jim19930609**
Issue: https://github.com/taichi-dev/taichi/issues/5819

### CODE REVIEW

**Comment by taichi-gardener** [comment]
Rebase failed, please resolve the conflicts manually.

File: python/taichi/lang/ops.py [hunk 1]
```diff
@@ -7,7 +7,26 @@
...
+def uniform_matrix_inputs(*args):
+    has_real_matrix = False
+    for arg in args:
+        if is_taichi_expr(arg) and arg.ptr.is_tensor():
+            has_real_matrix = True
+            break
+
+    results = []
+    for arg in args:
+        if has_real_matrix and is_matrix_class(arg):
```
Comment (strongoier) [inline python/taichi/lang/ops.py:23]: Can we simply use `isinstance(arg, Matrix)` instead of adding `is_matrix_class`?

Comment (jim19930609) [reply python/taichi/lang/ops.py:23]: Interesting, the root reason is that Python doesn't support separate declaration and definition, therefore circular import can be a real headache. 

However, if it supports separate declaration and definition, we'll then suffer from symbol inconsistency instead :(

**Pull request merged**